api4/hosted_customer.go

```diff
@@ -71,3 +89,176 @@ func selfHostedBootstrap(c *Context, w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	w.Write(json)
+}
+
+func selfHostedConfirm(c *Context, w http.ResponseWriter, r *http.Request) {
+	const where = "Api4.selfHostedConfirm"
+	ensureSelfHostedAdmin(c, where)
+	if c.Err != nil {
+		return
+	}
+	if !checkSelfHostedPurchaseEnabled(c) {
+		c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	bodyBytes, err := io.ReadAll(r.Body)
+	if err != nil {
+		c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusBadRequest).Wrap(err)
+		return
+	}
+
+	var confirm model.SelfHostedConfirmPaymentMethodRequest
+	err = json.Unmarshal(bodyBytes, &confirm)
+	if err != nil {
+		c.Err = model.NewAppError(where, "api.cloud.request_error", nil, "", http.StatusBadRequest).Wrap(err)
+		return
+	}
+
+	user, userErr := c.App.GetUser(c.AppContext.Session().UserId)
+	if userErr != nil {
+		c.Err = userErr
+		return
+	}
+	confirmResponse, err := c.App.Cloud().ConfirmSelfHostedSignup(confirm, user.Email)
+	if err != nil {
+		if err.Error() == fmt.Sprintf("%d", http.StatusUnprocessableEntity) {
+			c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusUnprocessableEntity).Wrap(err)
+			return
+		}
+		c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		return
+	}
+	license, err := c.App.Srv().Platform().SaveLicense([]byte(confirmResponse.License))
+	// dealing with an AppError
+	if !(reflect.ValueOf(err).Kind() == reflect.Ptr && reflect.ValueOf(err).IsNil()) {
+		c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		return
+	}
+	clientResponse, err := json.Marshal(model.SelfHostedSignupConfirmClientResponse{
+		License:  utils.GetClientLicense(license),
+		Progress: confirmResponse.Progress,
+	})
+	if err != nil {
+		c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		return
+	}
+
+	go func() {
+		err := c.App.Cloud().ConfirmSelfHostedSignupLicenseApplication()
+		if err != nil {
+			c.Logger.Warn("Unable to confirm license application", mlog.Err(err))
+		}
+	}()
+
+	_, _ = w.Write(clientResponse)
+}
+
+func handleSignupAvailable(c *Context, w http.ResponseWriter, r *http.Request) {
+	const where = "Api4.handleSignupAvailable"
+	ensureSelfHostedAdmin(c, where)
+	if c.Err != nil {
+		return
+	}
+	if !checkSelfHostedPurchaseEnabled(c) {
+		c.Err = model.NewAppError(where, "api.cloud.app_error", nil, "", http.StatusNotImplemented)
+		return
+	}
+	if err := c.App.Cloud().SelfHostedSignupAvailable(); err != nil {
+		c.Err = model.NewAppError(where, "api.server.hosted_signup_unavailable.error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	ReturnStatusOK(w)
+}
+
+func selfHostedInvoices(c *Context, w http.ResponseWriter, r *http.Request) {
+	const where = "Api4.selfHostedInvoices"
+	ensureSelfHostedAdmin(c, where)
```

Comment:
Since self-hosted starter means enterprise edition with no license uploaded and it's also used in the bootstrap process, shouldn't we have a nil check for license within this function?

Reply:
The `license.IsCloud()` helper is safe for use against nil licenses. This test confirms that: https://github.com/mattermost/mattermost-server/blob/master/model/license_test.go#L162-L165